python/pylibraft/pylibraft/neighbors/refine.pyx

```diff
@@ -0,0 +1,461 @@
+    return (shape, dtype, data)
+
+
+cdef host_matrix_view[float, uint64_t, row_major] \
+        get_host_matrix_view_float(array) except *:
+    shape, dtype, data = _get_array_params(
+        array.__array_interface__, check_dtype=np.float32)
+    return make_host_matrix_view[float, uint64_t, row_major](
+        <float*><uintptr_t>data, shape[0], shape[1])
+
+
+cdef host_matrix_view[uint64_t, uint64_t, row_major] \
+        get_host_matrix_view_uint64(array) except *:
+    shape, dtype, data = _get_array_params(
+        array.__array_interface__, check_dtype=np.uint64)
+    return make_host_matrix_view[uint64_t, uint64_t, row_major](
+        <uint64_t*><uintptr_t>data, shape[0], shape[1])
+
+
+cdef host_matrix_view[uint8_t, uint64_t, row_major] \
+        get_host_matrix_view_uint8(array) except *:
+    shape, dtype, data = _get_array_params(
+        array.__array_interface__, check_dtype=np.uint8)
+    return make_host_matrix_view[uint8_t, uint64_t, row_major](
+        <uint8_t*><uintptr_t>data, shape[0], shape[1])
+
+
+cdef host_matrix_view[int8_t, uint64_t, row_major] \
+        get_host_matrix_view_int8(array) except *:
+    shape, dtype, data = _get_array_params(
+        array.__array_interface__, check_dtype=np.int8)
+    return make_host_matrix_view[int8_t, uint64_t, row_major](
+        <int8_t*><uintptr_t>data, shape[0], shape[1])
+
+
+@auto_sync_handle
+def refine(dataset, queries, candidates, k=None, indices=None, distances=None,
+           metric="l2_expanded", handle=None):
+    """
+    Refine nearest neighbor search.
+
+    Refinement is an operation that follows an approximate NN search. The
+    approximate search has already selected n_candidates neighbor candidates
+    for each query. We narrow it down to k neighbors. For each query, we
+    calculate the exact distance between the query and its n_candidates
+    neighbor candidate, and select the k nearest ones.
+
+    Input arrays can be either CUDA array interface compliant matrices or
+    array interface compliant matrices in host memory. All array must be in
+    the same memory space.
+
+    Parameters
+    ----------
+    index_params : IndexParams object
+    dataset : array interface compliant matrix, shape (n_samples, dim)
+        Supported dtype [float, int8, uint8]
+    queries : array interface compliant matrix, shape (n_queries, dim)
+        Supported dtype [float, int8, uint8]
+    candidates : array interface compliant matrix, shape (n_queries, k0)
+        dtype [float, int8, uint64_t]
+    k : int
+        Number of neighbors to search (k <= k0). Optional if indices or
+        distances arrays are given (in which case their second dimension
+        is k).
+    indices :  Optional CUDA array interface compliant matrix shape
+                (n_queries, k), dtype uint64_t. If supplied, neighbor
+                indices will be written here in-place. (default None)
+        Supported dtype [float, int8, uint8]
```

Comment:
Should the supported dtypes be for the distances?

Reply:
I believe this belongs nor here (`uint64_t`) neither to distances (`float` only).